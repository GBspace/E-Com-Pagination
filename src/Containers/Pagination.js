import React from 'react';
import PropTypes from 'prop-types';

 
const propTypes = {
    items: PropTypes.array.isRequired,
    onChangePage: PropTypes.func.isRequired,
    pageSize: PropTypes.number.isRequired,
	initialPage: PropTypes.number
}
 
const defaultProps = {
    initialPage: 1,
	pageSize: 10
}

const isEqual = function (value, other) {
	let type = Object.prototype.toString.call(value);
	if (type !== Object.prototype.toString.call(other)) return false;
	if (['[object Array]', '[object Object]'].indexOf(type) < 0) return false;
	let valueLen = type === '[object Array]' ? value.length : Object.keys(value).length;
	let otherLen = type === '[object Array]' ? other.length : Object.keys(other).length;
	if (valueLen !== otherLen) return false;
	let compare = function (item1, item2) {
		let itemType = Object.prototype.toString.call(item1);
		if (['[object Array]', '[object Object]'].indexOf(itemType) >= 0) {
			if (!isEqual(item1, item2)) return false;
		}
		else {
			if (itemType !== Object.prototype.toString.call(item2)) return false;
			if (itemType === '[object Function]') {
				if (item1.toString() !== item2.toString()) return false;
			} else {
				if (item1 !== item2) return false;
			}
		}
	};
	if (type === '[object Array]') {
		for (let i = 0; i < valueLen; i++) {
			if (compare(value[i], other[i]) === false) return false;
		}
	} else {
		for (let key in value) {
			if (value.hasOwnProperty(key)) {
				if (compare(value[key], other[key]) === false) return false;
			}
		}
	}
	return true;
};
 
class Pagination extends React.Component {
    constructor(props) {
        super(props);
        this.state = { pager: {}};

		this.setPage=this.setPage.bind(this);
		this.getPager=this.getPager.bind(this);

	}
 
    componentWillMount=()=> {
        
        if (this.props.items && this.props.items.length) {
			this.setPage(this.props.initialPage);
        }
    }
 
    componentDidUpdate=(prevProps)=> {

        if (!(isEqual(prevProps.items,this.props.items)) || prevProps.pageSize !== this.props.pageSize ) {
				this.setPage(this.props.initialPage);
	        }
	}

    setPage=(page)=>{
	
        var { items, pageSize } = this.props;
		var pager = this.state.pager;
 
        if (page < 1 || page > pager.totalPages) {
            return;
        }
 
        pager = this.getPager(items.length, page, pageSize);
        var pageOfItems = items.slice(pager.startIndex, pager.endIndex + 1);
        this.setState({ pager: pager });
        this.props.onChangePage(pageOfItems);
    }
 
    getPager=(totalItems, currentPage, pageSize)=>{
		currentPage = currentPage || 1;
        pageSize = pageSize || 8;
		var totalPages = Math.ceil(totalItems / pageSize);
        var startPage, endPage;
        if (totalPages <= 10) {
            startPage = 1;
            endPage = totalPages;
        } else {
            
            if (currentPage <= 6) {
                startPage = 1;
                endPage = 10;
            } else if (currentPage + 4 >= totalPages) {
                startPage = totalPages - 9;
                endPage = totalPages;
            } else {
                startPage = currentPage - 5;
                endPage = currentPage + 4;
            }
        }
 
        const startIndex = (currentPage - 1) * pageSize;
        const endIndex = Math.min(startIndex + pageSize - 1, totalItems - 1);
 
        const pages = [...Array((endPage + 1) - startPage).keys()].map(i => startPage + i);
		return {
            totalItems: totalItems,
            currentPage: currentPage,
            pageSize: pageSize,
            totalPages: totalPages,
            startPage: startPage,
            endPage: endPage,
            startIndex: startIndex,
            endIndex: endIndex,
            pages: pages
        };
    }
 

    render() {
		
		let pager = this.state.pager;
		
        let activePage = pager.currentPage;
        if (!pager.pages || pager.pages.length <= 1) {
            return null;
        }
        return (
		
        <div>
            <ul className="list-group list-inline pull-right">
                <li className={pager.currentPage === 1 ? 'list-group-item disabled' : 'list-group-item'}
                     onClick={() => this.setPage(pager.currentPage - 1)}>Previous</li>
                <li key={activePage} className={pager.currentPage === activePage ? 'list-group-item active' : 'list-group-item'}
                    onClick={() => this.setPage(pager.currentPage)}>{activePage}</li>
                <li key={activePage+1} className={'list-group-item'}
                    onClick={() => this.setPage(pager.currentPage+1)}>{activePage+1}</li>
                <li key={activePage+2} className={'list-group-item'}
                    onClick={() => this.setPage(pager.currentPage+2)}>{activePage+2}</li>
                <li className={pager.currentPage === pager.totalPages ? 'list-group-item disabled' : 'list-group-item'}
                    onClick={() => this.setPage(pager.currentPage + 1)}>Next</li>
            </ul>
        </div>
        );
    }
}
 
Pagination.propTypes = propTypes;
Pagination.defaultProps = defaultProps;


export default Pagination;