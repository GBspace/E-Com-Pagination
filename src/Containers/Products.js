import React from 'react';
import {connect} from 'react-redux';
import {fetchProducts} from '../actions/Products';
import {getProducts,itemsPerPageCount} from '../selectors/Products';
import * as R from 'ramda';
import Pagination from './Pagination';



class Products extends React.Component{

    constructor(props){
        super(props);
        this.state={
            exampleItems: [],
            pageOfItems: [],
            pageSize:''
        };

        this.onChangePage = this.onChangePage.bind(this);
      
      
    };
    
    onChangePage=(pageOfItems)=>{
        this.setState({ pageOfItems: pageOfItems });
    }
    
  

    async componentDidMount(){
       this.props.fetchProducts();
    }


    componentWillReceiveProps=(nextProps)=>{
		if(this.props.pageSize !== nextProps.pageSize){
			this.setState({
				pageSize: nextProps.pageSize
            });
            
        }
	};

   renderProduct = (product,index)=>{
        const shortDesc = `${R.take(60,product.description)}...`;
        return (
            <div className='col-sm-3 col-lg-3 col-md-3 book-list' key={index}>
                <div className="thumbnail">
                   
                    <img className='img-thumbnail'
                        src={product.product_image}
                        alt={product.product_name}
                    />
                    <div className="caption">
                    <h5> {product.product_name} </h5>
                    <h6> {shortDesc}</h6>
                    <h4>
                        ${product.price}
                    </h4>
                 
                   
                    
                </div>
                </div>
                
            </div>
        );
    };

    render(){
        const {Products} = this.props;
        return(
        <div>
            <div className="books row">
                {
                    this.state.pageOfItems.map((product,index)=>{
                        return this.renderProduct(product,index);
                    })
                }
            </div>
            <div className="row">
                <div className="col-md-12">
                    <Pagination className=" pull-right"
                        items={Object.values(Products)}
                        onChangePage={this.onChangePage}
                        pageSize={this.state.pageSize}
                    >
                    </Pagination>
                </div>
            </div>
           
        </div>            
       
        )};
};

const mapDispatchToProps = (dispatch)=>{
    return{
    fetchProducts: ()=>dispatch(fetchProducts())
}};

const mapStateToProps = (state)=>({
        Products: getProducts(state),
        pageSize:parseInt(itemsPerPageCount(state)),

});

export default connect(mapStateToProps,mapDispatchToProps)(Products);

