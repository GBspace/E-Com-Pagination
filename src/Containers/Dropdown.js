import React from 'react';
import {connect} from 'react-redux';
import {updateItemsCount} from '../actions/Products';
import {itemsPerPageCount} from '../selectors/Products';

export class Dropdown extends React.Component{

    constructor(props){
        super(props);
        this.handleClick=this.handleClick.bind(this);
    };
    handleClick=(e)=>{
        e.preventDefault();
       
        this.props.updateItemsCount(e.target.id);
    };

    render(){
        return(
            <div className="dropdown pull-right" >
                <button className="btn btn-default dropdown-toggle" type="button" data-toggle="dropdown">{this.props.pageSize} Per Page
                <span className="caret"></span></button>
                <ul className="dropdown-menu dropdown-items">
                    <li><a id={8} onClick={this.handleClick}>8</a></li>
                    <li><a id={16} onClick={this.handleClick}>16</a></li>
                    <li><a id={24} onClick={this.handleClick}>24</a></li>
                </ul>
          </div>
        );
    };
};


const mapDispatchToProps=(dispatch)=>({
    updateItemsCount: count=>dispatch(updateItemsCount(count))
});

const mapStateToProps=(state)=>({
    pageSize:parseInt(itemsPerPageCount(state))
});

export default connect(mapStateToProps,mapDispatchToProps)(Dropdown);




