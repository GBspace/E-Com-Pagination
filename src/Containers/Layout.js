import React from 'react';
import Dropdown from './Dropdown';
import TotalProducts from './TotalProducts';
import Products from './Products';

export const Layout = ()=>{
    return (
        <div className="view-container">
            <div className="container">  
            
                <div className="row head-container">
                    <div className="col-sm-3 col-lg-3 col-md-3 pull-left">
                        <TotalProducts/> 
                    </div>
                    <div className="col-sm-3 col-lg-3 col-md-3 pull-right">
                        <Dropdown/>
                    </div>
                </div>
                
                <div className="row">
                    <div className="col-md-12">
                        <Products/>
                    </div>
                </div>
            </div>
            
        </div>
    )
};

export default Layout;