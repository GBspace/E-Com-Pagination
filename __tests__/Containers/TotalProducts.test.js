import React from 'react';
import {TotalProducts} from '../../src/Containers/TotalProducts';
import {shallow} from 'enzyme';

test('should render TotalProducts component correctly with value 100',()=>{
    const wrapper=shallow(<TotalProducts totalProd={10} />);
    expect(wrapper).toMatchSnapshot();
});