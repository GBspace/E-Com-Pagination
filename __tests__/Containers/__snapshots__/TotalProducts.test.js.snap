// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render TotalProducts component correctly with value 100 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TotalProducts
    totalProd={10}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h4>
           All Products
</h4>,
        <span
          className="label label-default"
>
          Total Products : 
          10
</span>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": " All Products",
        },
        "ref": null,
        "rendered": " All Products",
        "type": "h4",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "Total Products : ",
            10,
          ],
          "className": "label label-default",
        },
        "ref": null,
        "rendered": Array [
          "Total Products : ",
          10,
        ],
        "type": "span",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h4>
             All Products
</h4>,
          <span
            className="label label-default"
>
            Total Products : 
            10
</span>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": " All Products",
          },
          "ref": null,
          "rendered": " All Products",
          "type": "h4",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "Total Products : ",
              10,
            ],
            "className": "label label-default",
          },
          "ref": null,
          "rendered": Array [
            "Total Products : ",
            10,
          ],
          "type": "span",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
